src/posts/gpu-pipeline/index.md

```diff
@@ -0,0 +1,223 @@
+
+(TODO put in benchmark numbers here).
+
+For kvikio performance improvements, you need GPU Direct Storage (GDS) enabled on your system. This is a feature that allows the GPU to access data directly from storage, bypassing the CPU and reducing latency. GDS is supported on NVIDIA GPUs with the [GPUDirect Storage](https://docs.nvidia.com/datacenter/pgp/gds/index.html) feature.
+
+### Step 3: GPU-based decompression with nvCOMP :rocket:
+
+For a fully GPU-native workflow, we can let the GPU do all of the work!
+This includes reading the compressed data, decompression (using nvCOMP), any augmentation steps, to the ML model training.
+
+![image](https://hackmd.io/_uploads/rJtz66YCyl.png)
+
+Sending compressed instead of uncompressed data to the GPU means less data transfer overall, reducing I/O latency from storage to device.
+To unlock this, we would need zarr-python to support GPU-based decompression codecs, with one for Zstandard (Zstd) currently being implemented at https://github.com/zarr-developers/zarr-python/pull/2863.
+
+![image](https://hackmd.io/_uploads/H1kJkk9Akx.png)
+
+Figure above shows benchmark comparing CPU vs GPU-based decompression, with or without GDS enabled.
+
+Keep an eye on this space!
+
+### Step 4: Overlapping CPU and GPU compute with NVIDIA DALI :twisted_rightwards_arrows:
+
+Ideally, we want to minimize pauses where the device (GPU) is waiting for the host (CPU) or vice versa. This is one of the reasons we went with [NVIDIA DALI](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/index.html) that enables overlapping CPU and GPU computation.
+
+Our full codebase is available at https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus for reference.
+To start, there is a [zarr_DALI](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/tree/main/zarr_DALI) folder with short, contained examples of a DALI pipeline loading from Zarr.
+
+Next, look at the [zarr_ML_optimization](https://github.com/pangeo-data/ncar-hackathon-xarray-on-gpus/tree/main/zarr_ML_optimization) folder that contains an end-to-end example on how this DALI pipeline can be integrated into a Pytorch Dataloader and full training workflow.
+
+![image](https://hackmd.io/_uploads/H1YVp6tR1l.png)
+
+(TODO insert better nsight profiling figure than above showing overlapping CPU and GPU compute)
+
+## Going Forward
+
+This work is still ongoing, and we are continuing to explore ways to optimize data loading and processing for large-scale geospatial AI/ML workflows. We started this work during a 3-day hackathon, and we are excited to continue this work in the future. During the hackathon, we were able to make significant progress in optimizing data loading and processing for large-scale geospatial AI/ML workflows.
+
+We are continuing to explore the following areas:
+
+- GPU Direct Storage (GDS) for optimal performance
+- NVIDIA DALI
+- Work out how to use GDS when reading from cloud object store instead of on-prem disk.
+- etc
```

Comment:
Want to shout out that reading/writing Zarr shards with GPU buffers (thanks @maxrjones and @TomAugspurger!) at https://github.com/zarr-developers/zarr-python/pull/2978 was just merged, and could go in here or somewhere above, depending on when this blog post gets published.

Reply:
Next, we should test sharding on Lustre with and w/o GPU buffers. 🙈